File: flutter_app/wzam/lib/models/wzdx_models.dart
import 'package:json_annotation/json_annotation.dart';

part 'wzdx_models.g.dart';

enum WorkZoneType {
  @JsonValue('static')
  static,
  @JsonValue('moving')
  moving,
  @JsonValue('planned_moving_area')
  plannedMovingArea,
}

enum LocationMethod {
  @JsonValue('channel_device_method')
  channelDeviceMethod,
  @JsonValue('sign_method')
  signMethod,
  @JsonValue('junction_method')
  junctionMethod,
  @JsonValue('other')
  other,
  @JsonValue('unknown')
  unknown,
}

enum VehicleImpact {
  @JsonValue('all-lanes-closed')
  allLanesClosed,
  @JsonValue('some-lanes-closed')
  someLanesClosed,
  @JsonValue('all-lanes-open')
  allLanesOpen,
  @JsonValue('alternating-one-way')
  alternatingOneWay,
  @JsonValue('some-lanes-closed-merge-left')
  someLanesClosedMergeLeft,
  @JsonValue('some-lanes-closed-merge-right')
  someLanesClosedMergeRight,
  @JsonValue('all-lanes-open-shift-left')
  allLanesOpenShiftLeft,
  @JsonValue('all-lanes-open-shift-right')
  allLanesOpenShiftRight,
  @JsonValue('some-lanes-closed-split')
  someLanesClosedSplit,
  @JsonValue('flagging')
  flagging,
  @JsonValue('temporary-traffic-signal')
  temporaryTrafficSignal,
  @JsonValue('unknown')
  unknown,
}

enum LaneType {
  @JsonValue('exit-lane')
  exitLane,
  @JsonValue('exit-ramp')
  exitRamp,
  @JsonValue('entrance-lane')
  entranceLane,
  @JsonValue('entrance-ramp')
  entranceRamp,
  @JsonValue('sidewalk')
  sidewalk,
  @JsonValue('bike-lane')
  bikeLane,
  @JsonValue('shoulder')
  shoulder,
  @JsonValue('parking')
  parking,
  @JsonValue('median')
  median,
  @JsonValue('two-way-center-turn-lane')
  twoWayCenterTurnLane,
}

enum LaneStatus {
  @JsonValue('open')
  open,
  @JsonValue('closed')
  closed,
  @JsonValue('shift-left')
  shiftLeft,
  @JsonValue('shift-right')
  shiftRight,
  @JsonValue('merge-left')
  mergeLeft,
  @JsonValue('merge-right')
  mergeRight,
  @JsonValue('alternating-flow')
  alternatingFlow,
}

enum RestrictionType {
  @JsonValue('local-access-only')
  localAccessOnly,
  @JsonValue('no-trucks')
  noTrucks,
  @JsonValue('travel-peak-hours-only')
  travelPeakHoursOnly,
  @JsonValue('hov-3')
  hov3,
  @JsonValue('hov-2')
  hov2,
  @JsonValue('no-parking')
  noParking,
  @JsonValue('reduced-width')
  reducedWidth,
  @JsonValue('reduced-height')
  reducedHeight,
  @JsonValue('reduced-length')
  reducedLength,
  @JsonValue('reduced-weight')
  reducedWeight,
  @JsonValue('axle-load-limit')
  axleLoadLimit,
  @JsonValue('gross-weight-limit')
  grossWeightLimit,
  @JsonValue('towing-prohibited')
  towingProhibited,
  @JsonValue('permitted-oversize-loads-prohibited')
  permittedOversizeLoadsProhibited,
  @JsonValue('no-passing')
  noPassing,
}

enum UnitOfMeasurement {
  @JsonValue('feet')
  feet,
  @JsonValue('meters')
  meters,
  @JsonValue('centimeters')
  centimeters,
  @JsonValue('pounds')
  pounds,
  @JsonValue('tons')
  tons,
  @JsonValue('kilograms')
  kilograms,
}

@JsonSerializable()
class Restriction {
  final RestrictionType type;
  final double value;
  final UnitOfMeasurement unit;

  Restriction({
    required this.type,
    required this.value,
    required this.unit,
  });

  factory Restriction.fromJson(Map<String, dynamic> json) => _$RestrictionFromJson(json);
  Map<String, dynamic> toJson() => _$RestrictionToJson(this);
}

@JsonSerializable()
class Lane {
  @JsonKey(name: 'lane_order')
  final int laneOrder;
  final LaneType type;
  final LaneStatus status;
  final List<Restriction> restrictions;

  Lane({
    required this.laneOrder,
    required this.type,
    required this.status,
    required this.restrictions,
  });

  factory Lane.fromJson(Map<String, dynamic> json) => _$LaneFromJson(json);
  Map<String, dynamic> toJson() => _$LaneToJson(this);
}

enum WorkTypeName {
  @JsonValue('maintenance')
  maintenance,
  @JsonValue('minor-road-defect-repair')
  minorRoadDefectRepair,
  @JsonValue('roadside-work')
  roadsideWork,
  @JsonValue('overhead-work')
  overheadWork,
  @JsonValue('below-road-work')
  belowRoadWork,
  @JsonValue('barrier-work')
  barrierWork,
  @JsonValue('surface-work')
  surfaceWork,
  @JsonValue('painting')
  painting,
  @JsonValue('roadway-relocation')
  roadwayRelocation,
  @JsonValue('roadway-creation')
  roadwayCreation,
}

@JsonSerializable()
class TypeOfWork {
  @JsonKey(name: 'type_name')
  final WorkTypeName typeName;
  @JsonKey(name: 'is_architectural_change')
  final bool isArchitecturalChange;

  TypeOfWork({
    required this.typeName,
    required this.isArchitecturalChange,
  });

  factory TypeOfWork.fromJson(Map<String, dynamic> json) => _$TypeOfWorkFromJson(json);
  Map<String, dynamic> toJson() => _$TypeOfWorkToJson(this);
}

enum WorkerPresenceDefinition {
  @JsonValue('workers-in-work-zone-working')
  workersInWorkZoneWorking,
  @JsonValue('workers-in-work-zone-not-working')
  workersInWorkZoneNotWorking,
  @JsonValue('mobile-equipment-in-work-zone-moving')
  mobileEquipmentInWorkZoneMoving,
  @JsonValue('mobile-equipment-in-work-zone-not-moving')
  mobileEquipmentInWorkZoneNotMoving,
  @JsonValue('fixed-equipment-in-work-zone')
  fixedEquipmentInWorkZone,
  @JsonValue('humans-behind-barrier')
  humansBehindBarrier,
  @JsonValue('humans-in-right-of-way')
  humansInRightOfWay,
}

enum WorkerPresenceMethod {
  @JsonValue('camera-monitoring')
  cameraMonitoring,
  @JsonValue('arrow-board-present')
  arrowBoardPresent,
  @JsonValue('cones-present')
  conesPresent,
  @JsonValue('maintenance-vehicle-present')
  maintenanceVehiclePresent,
  @JsonValue('wearables-present')
  wearablesPresent,
  @JsonValue('mobile-device-present')
  mobileDevicePresent,
  @JsonValue('check-in-app')
  checkInApp,
  @JsonValue('check-in-verbal')
  checkInVerbal,
  @JsonValue('scheduled')
  scheduled,
}

enum WorkerPresenceConfidence {
  @JsonValue('low')
  low,
  @JsonValue('medium')
  medium,
  @JsonValue('high')
  high,
}

@JsonSerializable()
class WorkerPresence {
  @JsonKey(name: 'are_workers_present')
  final bool areWorkersPresent;
  final WorkerPresenceDefinition definition;
  final WorkerPresenceMethod method;
  @JsonKey(name: 'worker_presence_last_confirmed_date')
  final int workerPresenceLastConfirmedDate;
  final WorkerPresenceConfidence confidence;

  WorkerPresence({
    required this.areWorkersPresent,
    required this.definition,
    required this.method,
    required this.workerPresenceLastConfirmedDate,
    required this.confidence,
  });

  factory WorkerPresence.fromJson(Map<String, dynamic> json) => _$WorkerPresenceFromJson(json);
  Map<String, dynamic> toJson() => _$WorkerPresenceToJson(this);
}

File: api/app/wzdx_models.py
from enum import Enum
from pydantic import BaseModel


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/WorkZoneType.md
class WorkZoneType(str, Enum):
    static = "static"
    moving = "moving"
    planned_moving_area = "planned_moving_area"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/LocationMethod.md
class LocationMethod(str, Enum):
    channel_device_method = "channel_device_method"
    sign_method = "sign_method"
    junction_method = "junction_method"
    other = "other"
    unknown = "unknown"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/VehicleImpact.md
class VehicleImpact(str, Enum):
    all_lanes_closed = "all-lanes-closed"
    some_lanes_closed = "some-lanes-closed"
    all_lanes_open = "all-lanes-open"
    alternating_one_way = "alternating-one-way"
    some_lanes_closed_merge_left = "some-lanes-closed-merge-left"
    some_lanes_closed_merge_right = "some-lanes-closed-merge-right"
    all_lanes_open_shift_left = "all-lanes-open-shift-left"
    all_lanes_open_shift_right = "all-lanes-open-shift-right"
    some_lanes_closed_split = "some-lanes-closed-split"
    flagging = "flagging"
    temporary_traffic_signal = "temporary-traffic-signal"
    unknown = "unknown"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/LaneType.md
class LaneType(str, Enum):
    general = "general"
    exit_lane = "exit-lane"
    exit_ramp = "exit-ramp"
    entrance_lane = "entrance-lane"
    entrance_ramp = "entrance-ramp"
    sidewalk = "sidewalk"
    bike_lane = "bike-lane"
    shoulder = "shoulder"
    parking = "parking"
    median = "median"
    two_way_center_turn_lane = "two-way-center-turn-lane"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/LaneStatus.md
class LaneStatus(str, Enum):
    open = "open"
    closed = "closed"
    shift_left = "shift-left"
    shift_right = "shift-right"
    merge_left = "merge-left"
    merge_right = "merge-right"
    alternating_flow = "alternating-flow"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/RestrictionType.md
class RestrictionType(str, Enum):
    local_access_only = "local-access-only"
    no_trucks = "no-trucks"
    travel_peak_hours_only = "travel-peak-hours-only"
    hov_3 = "hov-3"
    hov_2 = "hov-2"
    no_parking = "no-parking"
    reduced_width = "reduced-width"
    reduced_height = "reduced-height"
    reduced_length = "reduced-length"
    reduced_weight = "reduced-weight"
    axle_load_limit = "axle-load-limit"
    gross_weight_limit = "gross-weight-limit"
    towing_prohibited = "towing-prohibited"
    permitted_oversize_loads_prohibited = "permitted-oversize-loads-prohibited"
    no_passing = "no-passing"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/UnitOfMeasurement.md
class UnitOfMeasuremenet(str, Enum):
    feet = "feet"
    meters = "meters"
    centimeters = "centimeters"
    poinds = "pounds"
    tons = "tons"
    kilograms = "kilograms"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/Restriction.md
class Restriction(BaseModel):
    type: RestrictionType
    value: float
    unit: UnitOfMeasuremenet


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/Lane.md
class Lane(BaseModel):
    lane_order: int
    type: LaneType
    status: LaneStatus
    restrictions: list[Restriction]


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/WorkTypeName.md
class WorkTypeName(str, Enum):
    maintenance = "maintenance"
    minor_road_defect_repair = "minor-road-defect-repair"
    roadside_work = "roadside-work"
    overhead_work = "overhead-work"
    below_road_work = "below-road-work"
    barrier_work = "barrier-work"
    surface_work = "surface-work"
    painting = "painting"
    roadway_relocation = "roadway-relocation"
    roadway_creation = "roadway-creation"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/TypeOfWork.md
class TypeOfWork(BaseModel):
    type_name: WorkTypeName
    is_architectural_change: bool


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/WorkerPresenceDefinition.md
class WorkerPresenceDefinition(str, Enum):
    workers_in_work_zone_working = "workers-in-work-zone-working"
    workers_in_work_zone_not_working = "workers-in-work-zone-not-working"
    mobile_equipment_in_work_zone_moving = "mobile-equipment-in-work-zone-moving"
    mobile_equipment_in_work_zone_not_moving = (
        "mobile-equipment-in-work-zone-not-moving"
    )
    fixed_equipment_in_work_zone = "fixed-equipment-in-work-zone"
    humans_behind_barrier = "humans-behind-barrier"
    humans_in_right_of_way = "humans-in-right-of-way"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/WorkerPresenceMethod.md
class WorkerPresenceMethod(str, Enum):
    camera_monitoring = "camera-monitoring"
    arrow_board_present = "arrow-board-present"
    cones_present = "cones-present"
    maintenance_vehicle_present = "maintenance-vehicle-present"
    wearables_present = "wearables-present"
    mobile_device_present = "mobile-device-present"
    check_in_app = "check-in-app"
    check_in_verbal = "check-in-verbal"
    scheduled = "scheduled"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/enumerated-types/WorkerPresenceConfidence.md
class WorkerPresenceConfidence(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


# https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/WorkerPresence.md
class WorkerPresence(BaseModel):
    are_workers_present: bool
    definition: WorkerPresenceDefinition
    method: WorkerPresenceMethod
    worker_presence_last_confirmed_date: int
    confidence: WorkerPresenceConfidence
